service/networking.py

```diff
@@ -1,8 +1,303 @@
+        return ExternalRouter
+    else:
+        raise Exception("Unknown topology name %s" % topology_name)
+
+
+def get_ranges(uid_number, inc=0):
+    """
+    Return two block ranges to be used to create subnets for
+    Atmosphere users.
+
+    NOTE: If you change MAX_SUBNET then you should likely change
+    the related math.
+    """
+    MAX_SUBNET = 4064  # Note 16 * 256
+    n = uid_number % MAX_SUBNET
+
+    #16-31
+    block1 = (n + inc) % 16 + 16
+
+    #1-254
+    block2 = ((n + inc) / 16) % 254 + 1
+
+    return (block1, block2)
+
+
+def get_default_subnet(username, inc=0, get_uid_number=None):
+    """
+    Return the default subnet for the username and provider.
+
+    Add and mod by inc to allow for collitions.
+    """
+    if get_uid_number:
+        uid_number = get_uid_number(username)
+    else:
+        uid_number = 0
+
+    if uid_number:
+        (block1, block2) = get_ranges(uid_number, inc)
+    else:
+        (block1, block2) = get_ranges(0, inc)
+
+    return "172.%s.%s.0/24" % (block1, block2)
+
+
+class NetworkTopology(object):
+    """
+    This network topology describes how networks should be created per openstack project.
+    """
+    name = None
+    user_network_required = False
+    user_router_required = False
+    options = {}
+
+    # Standard implementation for Network Topology
+    def __init__(self, identity):
+        pass
+
+    def configure(self, **options):
+        self.options.update(options)
+
+    def validate(self, core_identity):
+        """
+        Basic assertions, like 'username', 'project_name', 'password' could be added here...
+        """
+        return True
+
+    def get_or_create_network(self, network_driver, user_neutron, network_name):
+        network = network_driver.create_network(user_neutron, network_name)
+        return network
+
+    def get_or_create_user_subnet(
+            self, network_driver, neutron,
+            network_id, username, subnet_name,
+            ip_version=4, dns_nameservers=[],
+            get_unique_number=_get_unique_id,
+            get_cidr=get_default_subnet):
+        """
+        Create a subnet for the user using the get_cidr function to get
+        a private subnet range.
+        """
+        success = False
+        inc = 0
+        MAX_SUBNET = 4064
+        cidr = None
+        while not success and inc < MAX_SUBNET:
+            try:
+                cidr = get_cidr(username, inc, get_unique_number)
+                if cidr:
+                    return network_driver.create_subnet(neutron, subnet_name,
+                                              network_id, ip_version,
+                                              cidr, dns_nameservers)
+                else:
+                    logger.warn("Unable to create cidr for subnet "
+                                "for user: %s" % username)
+                    inc += 1
+            except NeutronClientException as nce:
+                if "overlap" in nce.message:
```

Comment:
I did a quick git grep of their code base. I could not find "overlap" in any of their raised exceptions. They raise specific exceptions, if there's a specific one, we should rely on that rather than parsing the message, which isn't an interface we can depend on.